phac-nml/irida-next
# Updatable User fields, local and omniauth

## What does this PR do and why?
Adds more fields to Users that can be populated by OmniAuth providers, or added by the User on the edit profile screen.

Fixes #72 

Adds the following parameters to User objects
* `first_name`
* `last_name`
* `phone_number`
* `provider_username`

Allow the following parameters on User objects to be updated via `update_params` and via `views/profiles/_form.html.erb`
* `email` - was already updatable
* `first_name`
* `last_name`
* `phone_number`

Added a alternative page to display user information for users that are managed by OmniAuth providers. This page is displayed instead of the user update form.
* `views/profiles/_domain_user.erb`

Added the new fields to the `developer` OmniAuth provider so all the fields can be accounted for and tested without needing to sign in with saml/azure.

## Screenshots or screen recordings
New user profile screen for local users
![image](https://github.com/phac-nml/irida-next/assets/16961365/a58c12a0-47ab-40fc-a84d-c1ebc5415b0d)

New user profile screen for OmniAuth users
![image](https://github.com/phac-nml/irida-next/assets/16961365/678337d6-badb-4b78-a07c-9dd4d1e2704a)

New register new user screen
![image](https://github.com/phac-nml/irida-next/assets/16961365/fd9a2f62-698b-46f0-b95c-c2d9c2b0ef00)


## How to set up and validate locally
*Note: screenshots below here are from before PR and UI design was updated, but steps are still valid*

1. run `bundle` and `rake db:drop db:create db:migrate`
2. Run IRIDA Next with the `developer` OmniAuth Provider enabled.
- This can be done by adding the following as an Environment Variable `"OMNIAUTH_PROVIDERS": "developer"`
3. Create a new user and sign in
4. Click edit profile 
![image](https://github.com/phac-nml/irida-next/assets/16961365/f06697ea-fe35-4f43-b0cd-0439a00131ed)
5. See that the new edit screen is viewable
![image](https://github.com/phac-nml/irida-next/assets/16961365/719a9d06-d71d-4d86-89c8-d94bf70d5a8c)
6. Verify the state of the User via rails console 
![image](https://github.com/phac-nml/irida-next/assets/16961365/a4c9aa03-5199-45bb-84f8-e43369971929)
7. fill in the form and add some new data 
![image](https://github.com/phac-nml/irida-next/assets/16961365/c3953f59-9a70-41c3-9738-e4edfe068431)
8. Verify the data has been updated
![image](https://github.com/phac-nml/irida-next/assets/16961365/7f516e7c-98a7-447f-8587-42cf2b302c30)
9. Sign out and sign in via Developer OmniAuth 
![image](https://github.com/phac-nml/irida-next/assets/16961365/a5772d87-46b6-4e04-b221-2ea81a9ac461)
10. Fill in the fields and Sign In. These mimic how SAML/Azure send and format data that will populate a user
![image](https://github.com/phac-nml/irida-next/assets/16961365/59fcdfa0-e01a-46b7-b00a-4d863f0aa87c)
11. Click edit profile 
![image](https://github.com/phac-nml/irida-next/assets/16961365/b291d55f-14e9-4acc-909d-c36602cb9d0e)
12. View the user details are being displayed 
![image](https://github.com/phac-nml/irida-next/assets/16961365/fe92c373-7f0f-4556-94eb-72ff43b765dd)
13. Verify the data via console 
![image](https://github.com/phac-nml/irida-next/assets/16961365/52710401-4bc3-44c9-99c0-765fe352b67c)
14. Sign out, and sign back in with some changed fields. the `email` field is configured to be the `uid` for the `developer` omniauth provider. You can see this configuration in `devise.rb`. So the `email`/`uid` must remain the same to update a users parameters when signing in with the `developer` provider. Different providers user different fields for `uid`. When signing in with omniauth using an existing `uid`, the user will be updated if the fields have changed, instead of a new user being created.
![image](https://github.com/phac-nml/irida-next/assets/16961365/3e32f28b-c5d8-49d1-8959-2f5d21acbd61)
15. Verify the data has been updated
![image](https://github.com/phac-nml/irida-next/assets/16961365/d3a55fa4-102e-4810-8ffb-0836b680a449)



## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

  the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.
